### ISSUE TITLE
The function 'slice.IndexOf' is not thread-safe

### ISSUE THREAD

**Issue by SinuxLee** [duke-git/lancet#262]
I get this information when I execute this function on different coroutines.
```
fatal error: concurrent map writes

goroutine 1483221 [running]:
github.com/duke-git/lancet/v2/slice.IndexOf[...]({0xc0026e2ea0, 0x6, 0x6}, {0x0, 0x0})
        .../go/pkg/mod/github.com/duke-git/lancet/v2@v2.3.3/slice/slice.go:1196 +0x54f

```

**Comment by duke-git** [comment]
> I get this information when I execute this function on different coroutines.
> 
> ```
> fatal error: concurrent map writes
> 
> goroutine 1483221 [running]:
> github.com/duke-git/lancet/v2/slice.IndexOf[...]({0xc0026e2ea0, 0x6, 0x6}, {0x0, 0x0})
>         .../go/pkg/mod/github.com/duke-git/lancet/v2@v2.3.3/slice/slice.go:1196 +0x54f
> ```

The methods of slice package are not goroutines safe and requires a lock to ensure goroutines safety.

**Comment by SinuxLee** [comment]
> > I get this information when I execute this function on different coroutines.
> > ```
> > fatal error: concurrent map writes
> > 
> > goroutine 1483221 [running]:
> > github.com/duke-git/lancet/v2/slice.IndexOf[...]({0xc0026e2ea0, 0x6, 0x6}, {0x0, 0x0})
> >         .../go/pkg/mod/github.com/duke-git/lancet/v2@v2.3.3/slice/slice.go:1196 +0x54f
> > ```
> 
> The methods of slice package are not goroutines safe and requires a lock to ensure goroutines safety.

Can we change memoryHashMap and memoryHashCounter to sync.map to support concurrency ?  Looking forward to your reply.

**Comment by duke-git** [comment]
> > > I get this information when I execute this function on different coroutines.
> > > ```
> > > fatal error: concurrent map writes
> > > 
> > > goroutine 1483221 [running]:
> > > github.com/duke-git/lancet/v2/slice.IndexOf[...]({0xc0026e2ea0, 0x6, 0x6}, {0x0, 0x0})
> > >         .../go/pkg/mod/github.com/duke-git/lancet/v2@v2.3.3/slice/slice.go:1196 +0x54f
> > > ```
> > 
> > 
> > The methods of slice package are not goroutines safe and requires a lock to ensure goroutines safety.
> 
> Can we change memoryHashMap and memoryHashCounter to sync.map to support concurrency ? Looking forward to your reply.

Okay, you can submit a PR to change it.

**Comment by cannian1** [comment]
> > > > I get this information when I execute this function on different coroutines.
> > > > ```
> > > > fatal error: concurrent map writes
> > > > 
> > > > goroutine 1483221 [running]:
> > > > github.com/duke-git/lancet/v2/slice.IndexOf[...]({0xc0026e2ea0, 0x6, 0x6}, {0x0, 0x0})
> > > >         .../go/pkg/mod/github.com/duke-git/lancet/v2@v2.3.3/slice/slice.go:1196 +0x54f
> > > > ```
> > > 
> > > 
> > > The methods of slice package are not goroutines safe and requires a lock to ensure goroutines safety.
> > 
> > 
> > Can we change memoryHashMap and memoryHashCounter to sync.map to support concurrency ? Looking forward to your reply.
> 
> Okay, you can submit a PR to change it.

I think we should evaluate the performance overhead of sync.Map and map + RWLock in here.

**Comment by duke-git** [comment]
@cannian1 make sense, we should evaluate the performance here.

**Comment by duke-git** [comment]
fixed in v2.3.4
